### PR TITLE
Allow items to skip IDs

### PR DESCRIPTION
**Pull request by silasary**
This is useful both for providing buffer space for future expansion (without shifting later IDs), and also just nice for organization reasons.

It's also an alternative to #73

### CODE REVIEW

**Comment by axxroytovu** [comment]
Can you add a syntax example to the `Items.json` file?

**Comment by nicopop** [comment]
1. this probably need to be added to the json schemas
2. might cause id issue down the line if an id is specified and later the automated count get counted up to it

**Comment by silasary** [comment]
> 1. this probably need to be added to the json schemas

Yup, can do.

> 2. might cause id issue down the line if an id is specified and later the automated count get counted up to it

This is why I throw an error if that happens.

File: src/Items.py [hunk 1]
```diff
@@ -25,7 +25,14 @@
...
 for key, val in enumerate(item_table):
-    item_table[key]["id"] = count
+    if "id" in item_table[key]:
+        item_id = item_table[key]["id"]
+        if item_id > count:
+            item_table[key]["id"] = item_id
+        else:
+            raise ValueError(f"{item_table[key]['name']} has an invalid ID. ID must be at least {count + 1}")
+    else:
+        item_table[key]["id"] = count
```
Comment (FuzzyGamesOn) [inline src/Items.py:35]: I think this breaks in a situation where world creators only set some of the item ids. Like, if I have these items:

- Item 1
- Item 2, id = 4
- Item 3
- Item 4

Then id 4 will be set for 2 items.

We might want to set count to the custom id when we encounter it. Then the increment starts counting from there.

File: src/Locations.py [hunk 2]
```diff
@@ -17,7 +17,15 @@
...
 
-    location_table[key]["id"] = count
+    if "id" in location_table[key]:
+        item_id = location_table[key]["id"]
+        if item_id > count:
+            location_table[key]["id"] = item_id
+        else:
+            raise ValueError(f"{location_table[key]['name']} has an invalid ID. ID must be at least {count + 1}")
+    else:
+        location_table[key]["id"] = count
```
Comment (FuzzyGamesOn) [inline src/Locations.py:27]: Same note as above.

**Review by FuzzyGamesOn — CHANGES_REQUESTED** [review]

File: src/Items.py (hunk 1 above)
Comment (nicopop) [reply src/Items.py:35]: > We might want to set count to the custom id when we encounter it. Then the increment starts counting from there.  

either that or we keep track of the custom ids somewhere and skip over them when counting up

Comment (silasary) [reply src/Items.py:35]: Updating count is absolutely the intended behaviour, thanks for catching that

**Review by nicopop — APPROVED** [review]

**Review by nicopop — DISMISSED** [review]
I dont think its as good as it could be since this wont allow the ID to go back at all but it should be good enough for now

File: src/Items.py [hunk 3]
```diff
@@ -25,6 +25,13 @@
 
 # add sequential generated ids to the lists
 for key, val in enumerate(item_table):
+    if "id" in item_table[key]:
+        item_id = item_table[key]["id"]
+        if item_id > count:
```
Comment (axxroytovu) [inline src/Items.py:30]: I don't think we should throw an error if `item_id == count` since that is just the default behavior.

Comment (axxroytovu) [reply src/Items.py:30]: Also, is there any way to account for someone not putting the item IDs in order, but listing valid numbers? E.g. [1, 2, blank, 4, 3] will break, but could be solved by sorting the items to be [1, 2, 3, 4, blank].

**Review by axxroytovu — CHANGES_REQUESTED** [review]

File: src/Items.py (hunk 3 above)
Comment (FuzzyGamesOn) [reply src/Items.py:30]: I'm not sure why someone would even want to have their custom ids out of order. That sounds like it should be something they should fix -- for their own sanity. 😜

Comment (silasary) [reply src/Items.py]: If we allow people to go backwards, it means we need to do collision detection on every single increment, and that's much more expensive than just asking people to have things in order

**Review by axxroytovu — APPROVED** [review]

**Review by FuzzyGamesOn — APPROVED** [review]

**Pull request merged**